File: fxgl/src/main/java/com/almasb/fxgl/extra/entity/state/State.java
/*
 * FXGL - JavaFX Game Library. The MIT License (MIT).
 * Copyright (c) AlmasB (almaslvl@gmail.com).
 * See LICENSE for details.
 */

package com.almasb.fxgl.extra.entity.state;

/**
 * @author Almas Baimagambetov (almaslvl@gmail.com)
 */
public abstract class State {

    /**
     * Called after entering this state from prevState
     */
    protected void onEnter(State prevState) {

    }

    protected abstract void onUpdate(double tpf);

    /**
     * Called before exit.
     */
    protected void onExit() {

    }
}

File: fxgl-core/src/main/kotlin/com/almasb/fxgl/core/fsm/State.kt
/*
 * FXGL - JavaFX Game Library. The MIT License (MIT).
 * Copyright (c) AlmasB (almaslvl@gmail.com).
 * See LICENSE for details.
 */

package com.almasb.fxgl.core.fsm

/**
 *
 * @author Almas Baimagambetov (almaslvl@gmail.com)
 */
interface State<T> {

    val isSubState: Boolean

    /**
     * Only honored if [isSubState] returns true.
     * Allow this substate to be active together with (sub)state above in the state machine
     * hierarchy.
     */
    val isAllowConcurrency: Boolean

    /**
     * Called before this state is to be set active and this state does not form
     * a part of the hierarchy. If it does, only [onEnter] is called.
     */
    fun onCreate()

    /**
     * Called after this state is removed from the hierarchy.
     */
    fun onDestroy()

    /**
     * Called after transitioning from [prevState].
     */
    fun onEnteredFrom(prevState: T)

    /**
     * Called before transitioning to [nextState].
     */
    fun onExitingTo(nextState: T)
}